lib/java/com/google/android/material/drawable/DrawableUtils.java

```diff
@@ -190,63 +194,124 @@ private static Drawable createTintableMutatedDrawableIfNeeded(
+
+    // Centers the top layer.
+    if (VERSION.SDK_INT >= VERSION_CODES.M) {
+      drawable.setLayerGravity(1, Gravity.CENTER);
     } else {
-      float topLayerRatio =
-          (float) topLayerDrawable.getIntrinsicWidth() / topLayerDrawable.getIntrinsicHeight();
-      float bottomLayerRatio =
-          (float) bottomLayerDrawable.getIntrinsicWidth()
-              / bottomLayerDrawable.getIntrinsicHeight();
+      // On Pre-M there's no layer gravity support, so we use layer insets to adjust the top layer
+      // position.
+      final int topLayerIntrinsicWidth = topLayerDrawable.getIntrinsicWidth();
+      final int topLayerIntrinsicHeight = topLayerDrawable.getIntrinsicHeight();
+      final int bottomLayerIntrinsicWidth = bottomLayerDrawable.getIntrinsicWidth();
+      final int bottomLayerIntrinsicHeight = bottomLayerDrawable.getIntrinsicHeight();
+
+      final int horizontalInset = max((bottomLayerIntrinsicWidth - topLayerIntrinsicWidth) / 2, 0);
+      final int verticalInset = max((bottomLayerIntrinsicHeight - topLayerIntrinsicHeight) / 2, 0);
+      drawable.setLayerInset(1,
+          horizontalInset, verticalInset, horizontalInset, verticalInset);
+    }
+
+    return drawable;
+  }
+
+  /**
+   * Composites two drawables, returning a drawable instance of {@link LayerDrawable},
+   * with the top layer centered and scaled with the aspect ratio preserved to the size of
+   * the bottom layer.
+   * <p>
+   * If any of the drawables is null, this method will return the other.
+   *
+   * @param bottomLayerDrawable the drawable to be on the bottom layer
+   * @param topLayerDrawable the drawable to be on the top layer
+   * @param topLayerDesiredWidth top layer desired width in pixels, or -1 to use the intrinsic
+   *     width if possible.
+   * @param topLayerDesiredHeight top layer desired height in pixels, or -1 to use the intrinsic
+   *     height if possible.
+   */
+  @RequiresApi(VERSION_CODES.M)
+  @Nullable
+  public static Drawable compositeDrawables(
+      @Nullable Drawable bottomLayerDrawable, @Nullable Drawable topLayerDrawable,
+      @Px int topLayerDesiredWidth, @Px int topLayerDesiredHeight) {
+
+    if (bottomLayerDrawable == null) {
+      return topLayerDrawable;
+    }
+
+    if (topLayerDrawable == null) {
+      return bottomLayerDrawable;
+    }
+
+    final int topLayerIntrinsicWidth = topLayerDrawable.getIntrinsicWidth();
+    final int topLayerIntrinsicHeight = topLayerDrawable.getIntrinsicHeight();
+    final int bottomLayerIntrinsicWidth = bottomLayerDrawable.getIntrinsicWidth();
+    final int bottomLayerIntrinsicHeight = bottomLayerDrawable.getIntrinsicHeight();
+
+    int topLayerNewWidth =
+        topLayerDesiredWidth != -1 ? topLayerDesiredWidth : topLayerIntrinsicWidth;
+    int topLayerNewHeight =
+        topLayerDesiredHeight != -1 ? topLayerDesiredHeight : topLayerIntrinsicHeight;
+
+    // If there's no intrinsic width or height, keep bottom layer's size.
```

Comment:
As I mentioned before, let's keep the original logic as much as possible. 

We should only override topLayerIntrinsicWidth / topLayerIntrinsicHeight with desired width / height in this change and keep other logic intact in this class. 

Unless - if we think it's meaningful to always respect desired width / height - even if they are larger than the bottom layer. In that case we can rethink a more clear structure to do that.

Reply:
> As I mentioned before, let's keep the original logic as much as possible.
>
> We should only override topLayerIntrinsicWidth / topLayerIntrinsicHeight with desired width / height in this change and keep other logic intact in this class.

It doesn't work. On pre-M, desired sizes will be ignored because pre-M `LayerDrawable` doesn't support scaling, and you will end up with a drawable that is both incorrectly sized and incorrectly centered.

Reply:
oh ok. I see what you mean. Maybe let me download the PR and play with it by myself for a little bit.